src/delphi/train/architectures.py

```diff
@@ -0,0 +1,73 @@
+from dataclasses import fields
+
+import torch
+from llama2c import model_export
+from llama2c.model import ModelArgs as Llama2ModelArgs
+from llama2c.model import Transformer as Llama2Model
+
+
+class ModelTypes:
+    LLAMA2C = "llama2c"
+    MAMBA = "mamba"
+
+
+args_to_load_from_checkpoint = {
+    ModelTypes.LLAMA2C: [
+        "dim",
+        "n_layers",
+        "n_heads",
+        "n_kv_heads",
+        "vocab_size",
+        "multiple_of",
+        "max_seq_len",
+    ],
+    ModelTypes.MAMBA: [
+        "n_layers",
+        "model_dim",
+    ],
+}
+
+
+def initialize_model(**model_args) -> torch.nn.Module:
+    if model_args["architecture"] == ModelTypes.LLAMA2C:
+        # filter model_args for fields in Llama2ModelArgs
+        llama2_arg_names = {f.name for f in fields(Llama2ModelArgs)}
+        llama2_args = {k: v for k, v in model_args.items() if k in llama2_arg_names}
+        return Llama2Model(Llama2ModelArgs(**llama2_args))
+    else:
+        raise NotImplementedError(
+            f"Architecture {model_args['architecture']} not yet implemented"
+        )
+
+
+def load_model(model_args, checkpoint) -> torch.nn.Module:
+    arch = model_args["architecture"]
+    checkpoint_model_args = checkpoint["model_args"]
+    for k in args_to_load_from_checkpoint[arch]:
+        model_args[k] = checkpoint_model_args[k]
+    if arch == ModelTypes.LLAMA2C:
+        # create the model
+        gptconf = Llama2ModelArgs(**model_args)
+        model = Llama2Model(gptconf)
+        state_dict = checkpoint["model"]
+        # fix the keys of the state dictionary :(
```

Comment:
Are we sure this still happens?

Reply:
No idea tbqh, that's just something I copied and never checked.